### ISSUE TITLE
select copy and paste in automation track

### ISSUE THREAD

**Issue by eland1** [LMMS/lmms#5157]
it would be amazing if you could select, copy and paste in the automation track. it is also nice if you  can place two points right under eachother. i think it becomes a sidechain dream.

**Comment by Spekular** [comment]
Selection is a duplicate of of #4602.

**Comment by musikBear** [comment]
> place two points right under eachother. 

@eland1 In respect to that, you can just set **Q** -quantification to highest value. You will never be able to hear the difference between 'same value' and a distance of 1/192. That is not audible

**Comment by CYBERDEViLNL** [comment]
I was looking into implementing this and found that most of the code is already there, it needs some modifications and some extra code. Is anyone working on this? Else I will give it a try :-)

**Comment by BaraMGB** [comment]
@CYBERDEVILNL  Actually I am on it. At first I want to implement the ability to have more than one control point per time key. I suppose you want to implement select, copy paste?

**Comment by CYBERDEViLNL** [comment]
@BaraMGB all right that's awesome! Yes I was talking about the select, copy and paste. Let me know if I can help :-)

**Comment by tecknixia** [comment]
I've been rewriting the mouse functions for clarity and I decided I might as well look into this, so in order to figure out how it works, I wrote down something like a map of the process that's involved.  I figured it might help.

```
AutomationEditorWindow::AutomationEditorWindow

	DropToolBar *editActionsToolBar = addDropToolBarToTop(tr("Edit actions"));
	ActionGroup* editModeGroup = new ActionGroup(this);

	QAction* drawAction = editModeGroup->addAction(embed::getIconPixmap("edit_draw"),
		tr("Draw mode (Shift+D)"));
	QAction* eraseAction = editModeGroup->addAction(embed::getIconPixmap("edit_erase"),
		tr("Erase mode (Shift+E)"));

	// these next two lines are in a different format, so I changed them to match the others
	QAction* selectAction = editModeGroup->addAction(embed::getIconPixmap("edit_select"),
		tr("Select: Disabled"));
	QAction* moveAction = editModeGroup->addAction(embed::getIconPixmap("edit_move"),
		tr("Move: Disabled"));
```
That is just defining them, so order doesn't matter (the above)

Here is my map (below):
```
	connect(editModeGroup, SIGNAL(triggered(int)), m_editor, SLOT(setEditMode(int)));
		-> QAction
			-> AutomationEditor::setEditMode(int mode)
				-> setEditMode((AutomationEditor::EditModes) mode);
					-> AutomationEditor::EditModes (enum) DRAW, ERASE, SELECT, MOVE (in header file)
						- the order of EditModes in header file determines which action is triggered
						- if you put 'erase' in front of 'draw', the draw button will erase, and erase button will draw
							-> m_editMode
        // there's two of them?
			-> AutomationEditor::setEditMode(AutomationEditor::EditModes mode)
				-> m_editMode = mode;
				-> AutomationEditor::removeSelection
				-> AutomationEditor::getSelectedValues
```
It's confusing to me, but essentially it seems to me that it just takes the signal of the buttons on the toolbar and puts `m_editMode` into the associated mode.
```
	editActionsToolBar->addAction(drawAction);
	editActionsToolBar->addAction(eraseAction);
	editActionsToolBar->addAction(selectAction);
	editActionsToolBar->addAction(moveAction);
```
The order of these (above) determines the order of the buttons on the toolbar.
The order does not affect the functionality
```
switch( m_editMode )
{
	case DRAW:
		if( m_mouseDownRight )
		{
			cursor = s_toolErase;
		}
		else if( m_action == MOVE_VALUE )
		{
			cursor = s_toolMove;
		}
		else
		{
			cursor = s_toolDraw;
		}

		break;
	case ERASE: cursor = s_toolErase; break;
	case SELECT: cursor = s_toolSelect; break;
	case MOVE: cursor = s_toolMove; break;
}
QPoint mousePosition = mapFromGlobal( QCursor::pos() );
	if( cursor != NULL && mousePosition.y() > TOP_MARGIN + SCROLLBAR_SIZE)
	{
		p.drawPixmap( mousePosition + QPoint( 8, 8 ), *cursor );
	}
```
And that just changes the cursor according to the mode.

From there my guess it's it's just a matter of `m_editMode` states within the code.

I'm thinking about combining a few Automation Editor issues, as I've talked about here:

**Automation: points deleted while moving other points can be restored abnormally**
#4671

**Comment by Spekular** [comment]
Personally I think this should be two separate issues, one for group selection and one for placing points at the same time. However, given the huge number of open issues I believe they can stay as one for now. Perhaps if one part is implemented and the other remains, the remaining one can be split into a new issue and this can be closed.

**Comment by tecknixia** [comment]
In the branch I'm working on [above] I've managed to get select and move working, but I'm running into three issues that seem to be related.

1. Issue #5254  - Select All - only selects the length of measures in Song Editor

![automationSelectAll](https://user-images.githubusercontent.com/50790262/67161419-f5823e00-f31f-11e9-85e7-0058584c4014.png)

2. Issue #4671  - [Maybe related:] while moving a selection, some points randomly disappear.

I'm looking into this.

3. Issue #??? - actually I couldn't find an issue for this one... but the timeline position bar is broken.

This will be needed for pasting into the editor at the correct position.  It doesn't move when play is pressed, so I imagine the code wasn't completed.  I'll be looking into this too.

**Comment by tecknixia** [comment]
@Spekular 

The more I play with this issue of moving a selection, the more it seems like the random points are disappearing as they move to a position on top of one another.  When I read what you said, I agreed about the separate issues, but then I noticed this.  It seems like the issues may be related after all.

**Comment by Spekular** [comment]
If two land on each other at least one will be deleted because we currently
don't support several points on top of each other. Points at the same
position is still a separate issue IMO, as you can implement (and benefit
from) one without the other (and this applies both ways).

**Comment by tecknixia** [comment]
@Spekular 

After thinking about it some more, if points are getting deleted because they're on top of each other, than allowing them to co-exist on the same tick would probably fix that issue.

However, this might involve some strategy in how the mouse functions run.  Even if the issues are related, I think you're right, it seems to deserve it's own issue because of the depth of the issue.  This issue can then be mentioned by that one.

**Comment by tecknixia** [comment]
Just realized... if it works when dragging a single point, why can't it work for multiple points?

I'll have to take a closer look, it seems likely possible to fix this without having to worry about the issue of two points occupying the same x position.

**Comment by tecknixia** [comment]
![selectionBoxMove](https://user-images.githubusercontent.com/50790262/67172881-ae796480-f382-11e9-8a13-d26a768f80dd.png)

- If m_selectedTick is greater than zero
- and if m_selectStartTick plus ticks_diff is less than zero
- then set ticks_diff equal to the negative of m_selectStartTick
- else, if m_selectStartTick plus m_selectedTick plus ticks_diff is less than zero
- then set ticks_diff to the negative of m_selectStartTick plus m_selectedTick
- after that, increment m_selectStartTick by the amount of ticks_diff

Um... what?!

Let me try again in normal language.

- if the difference in ticks between the tick the mouse moved to and the start of the selection is greater than zero
- and if the start click of the selection plus the number of ticks the selection was moved is less than zero
- then set the variable ticks_diff to the negative of the start tick of the selection area
- else, if the sum of the select box start tick, the length of the ticks between the tick at the start of the selection area and the current tick the mouse moved to, and the length between the tick at the start of moving the selection and the current tick the mouse moved to is less than zero
- then set the variable ticks_diff to the negative of the sum of the tick at the start of the selection area, and the amount of ticks between the start of the selection area and the current tick the mouse moved to

Um... what?!

I'm confused.

**Comment by tecknixia** [comment]
I had to redo some things, so I created a new branch for this with cleaner commits.
https://github.com/tecknixia/lmms/commits/autoSelectCopyPaste

- Select, move, cut, and copy all seem to be working now
- Fixed broken toolbar buttons, including right-click mode switching
- Made selection box transparent
- Enabled the ability to scroll while selecting
- Enabled "select all" but runs into an issue with being dependent on the song editor (#5254 )
- Paste, I believe, runs into issues with the time line position indicator (and also 5254)
- Automation point adjustment with wheel scroll and double click included (#5291 and #5292 )
- Point adjustment PRs also contain a fix for #3301 
- Cancel dragging by click right mouse button, kinda partially addresses #4671

I'll probably also be integrating a fix for #4040 and potentially looking into a few other Automation Editor issues.

#4877

**Comment by tecknixia** [comment]
I believe I made a little more progress with being able to drag multiple points (without deleting the others).  I think I know what needs to be done now.

- `mouseMoveEvent` (src/gui/editors/AutomationEditor.cpp) 
- `setDragValue` (src/core/AutomationPattern.cpp) 

The main problem is.. each time `setDragValue` is run: 
- it's only being passed the coordinates of one point
- it deletes the point at that time position
- takes "a picture" of the other points on the map
- sets `m_dragging` to true, which allows skipping of the above
- applies the picture of the points not being dragged (each time it runs) back onto the map
- then each time returns `this->putValue(time, value, quantPos, controlKey)`

For the basic idea of how to do this, it seems like the main thing that needs to be done is to take note of all the positions of the multiple points being moved, then delete those values from the map, and then copy the map of the others, so that you have a copy to place back onto the map each time a move is made.

I haven't yet looked at `putValue` in detail, might have to look at how to manage multiple points there too, but I figured it would be worth sharing this info.

I'm not going to be able to throw something together real quick, because I'm not yet that familiar with arrays, and objects, or whatever you might use in this situation with c++.  I'll have to do a little more studying first.
